// docusaurus-theme-github-codeblock/cjs/theme/ReferenceCodeBlock/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const CodeBlock_1 = __importDefault(require("@theme-init/CodeBlock"));
const client_1 = require("docusaurus-theme-github-codeblock/client");
const RunmeLink_js_1 = require("./RunmeLink.js");
const utils_js_1 = require("./utils.js");
const constants_js_1 = require("./constants.js");
const buttonBarStyles = {
    fontSize: '.9em',
    fontWeight: 600,
    color: '#0E75DD',
    textAlign: 'right',
    paddingBottom: '13px',
    textDecoration: 'underline'
};
function ReferenceCode(props) {
    const [fetchResultState, fetchResultStateDispatcher] = (0, react_1.useReducer)(utils_js_1.codeReducer, constants_js_1.initialFetchResultState);
    const codeSnippetDetails = (0, utils_js_1.parseReference)(props.children);
    if (fetchResultState.loading !== false) {
        (0, utils_js_1.fetchCode)(codeSnippetDetails, fetchResultStateDispatcher);
    }
    const titleMatch = props.metastring?.match(/title="(?<title>[^"]*)"/);
    const customProps = {
        ...props,
        metastring: titleMatch?.groups?.title
            ? `${props.metastring} title="${titleMatch?.groups?.title}"`
            : `${props.metastring} title="${codeSnippetDetails.title}"`,
        children: constants_js_1.initialFetchResultState.code,
    };
    const codeblockConfig = (0, client_1.useCodeblockThemeConfig)();
    const showButtons = codeblockConfig.showGithubLink || codeblockConfig.showRunmeLink;
    return (react_1.default.createElement("div", { className: 'docusaurus-theme-github-codeblock' },
        react_1.default.createElement(CodeBlock_1.default, { ...customProps }, fetchResultState.code),
        showButtons && (react_1.default.createElement("div", { style: buttonBarStyles },
            react_1.default.createElement(RunmeLink_js_1.RunmeLink, { reference: props.children, metastring: props.metastring }),
            codeblockConfig.showGithubLink && (react_1.default.createElement("a", { href: props.children, className: 'githubLink', style: constants_js_1.buttonStyles, target: "_blank" }, codeblockConfig.githubLinkLabel))))));
}
exports.default = ReferenceCode;
