// docusaurus-theme-github-codeblock/build/theme/ReferenceCodeBlock/index.js
import React, { useReducer } from 'react';
import CodeBlock from '@theme-init/CodeBlock';
import { useCodeblockThemeConfig } from 'docusaurus-theme-github-codeblock/client';
import { RunmeLink } from './RunmeLink.js';
import { parseReference, fetchCode, codeReducer } from './utils.js';
import { initialFetchResultState, buttonStyles } from './constants.js';
const buttonBarStyles = {
    fontSize: '.9em',
    fontWeight: 600,
    color: '#0E75DD',
    textAlign: 'right',
    paddingBottom: '13px',
    textDecoration: 'underline'
};
function ReferenceCode(props) {
    const [fetchResultState, fetchResultStateDispatcher] = useReducer(codeReducer, initialFetchResultState);
    const codeSnippetDetails = parseReference(props.children);
    if (fetchResultState.loading !== false) {
        fetchCode(codeSnippetDetails, fetchResultStateDispatcher);
    }
    const titleMatch = props.metastring?.match(/title="(?<title>[^"]*)"/);
    const customProps = {
        ...props,
        metastring: titleMatch?.groups?.title
            ? `${props.metastring} title="${titleMatch?.groups?.title}"`
            : `${props.metastring} title="${codeSnippetDetails.title}"`,
        children: initialFetchResultState.code,
    };
    const codeblockConfig = useCodeblockThemeConfig();
    const showButtons = codeblockConfig.showGithubLink || codeblockConfig.showRunmeLink;
    return (React.createElement("div", { className: 'docusaurus-theme-github-codeblock' },
        React.createElement(CodeBlock, { ...customProps }, fetchResultState.code),
        showButtons && (React.createElement("div", { style: buttonBarStyles },
            React.createElement(RunmeLink, { reference: props.children, metastring: props.metastring }),
            codeblockConfig.showGithubLink && (React.createElement("a", { href: props.children, className: 'githubLink', style: buttonStyles, target: "_blank" }, codeblockConfig.githubLinkLabel))))));
}
export default ReferenceCode;
