// @algolia/client-search/model/fetchedIndex.ts
// Code generated by OpenAPI Generator (https://openapi-generator.tech), manual changes will be lost - read more on https://github.com/algolia/api-clients-automation. DO NOT EDIT.

export type FetchedIndex = {
  /**
   * Index name.
   */
  name: string;

  /**
   * Index creation date. An empty string means that the index has no records.
   */
  createdAt: string;

  /**
   * Date and time when the object was updated, in RFC 3339 format.
   */
  updatedAt: string;

  /**
   * Number of records contained in the index.
   */
  entries: number;

  /**
   * Number of bytes of the index in minified format.
   */
  dataSize: number;

  /**
   * Number of bytes of the index binary file.
   */
  fileSize: number;

  /**
   * Last build time.
   */
  lastBuildTimeS: number;

  /**
   * Number of pending indexing operations. This value is deprecated and should not be used.
   */
  numberOfPendingTasks: number;

  /**
   * A boolean which says whether the index has pending tasks. This value is deprecated and should not be used.
   */
  pendingTask: boolean;

  /**
   * Only present if the index is a replica. Contains the name of the related primary index.
   */
  primary?: string;

  /**
   * Only present if the index is a primary index with replicas. Contains the names of all linked replicas.
   */
  replicas?: string[];

  /**
   * Only present if the index is a [virtual replica](https://www.algolia.com/doc/guides/managing-results/refine-results/sorting/how-to/sort-an-index-alphabetically/#virtual-replicas).
   */
  virtual?: boolean;
};
