// @algolia/client-search/model/log.ts
// Code generated by OpenAPI Generator (https://openapi-generator.tech), manual changes will be lost - read more on https://github.com/algolia/api-clients-automation. DO NOT EDIT.

import type { LogQuery } from './logQuery';

export type Log = {
  /**
   * Date and time of the API request, in RFC 3339 format.
   */
  timestamp: string;

  /**
   * HTTP method of the request.
   */
  method: string;

  /**
   * HTTP status code of the response.
   */
  answer_code: string;

  /**
   * Request body.
   */
  query_body: string;

  /**
   * Response body.
   */
  answer: string;

  /**
   * URL of the API endpoint.
   */
  url: string;

  /**
   * IP address of the client that performed the request.
   */
  ip: string;

  /**
   * Request headers (API keys are obfuscated).
   */
  query_headers: string;

  /**
   * SHA1 signature of the log entry.
   */
  sha1: string;

  /**
   * Number of API requests.
   */
  nb_api_calls: string;

  /**
   * Processing time for the query in milliseconds. This doesn\'t include latency due to the network.
   */
  processing_time_ms: string;

  /**
   * Index targeted by the query.
   */
  index?: string;

  /**
   * Query parameters sent with the request.
   */
  query_params?: string;

  /**
   * Number of search results (hits) returned for the query.
   */
  query_nb_hits?: string;

  /**
   * Queries performed for the given request.
   */
  inner_queries?: LogQuery[];
};
